gcp/modules/k8s-snapshots/service_account_iam.tf

```diff
@@ -0,0 +1,24 @@
+data "google_service_account" "gke_cluster_node" {
+  account_id = "gke-cluster-node"
+  project    = "${var.project_id}"
+}
+
+data "google_service_account" "gke_cluster_pod_k8s_snapshots" {
+  account_id = "gke-cluster-pod-k8s-snapshots"
+  project    = "${var.project_id}"
+}
+
+data "google_iam_policy" "pod_k8s_snapshots" {
+  binding {
+    role = "roles/iam.serviceAccountTokenCreator"
+
+    members = [
+      "serviceAccount:${data.google_service_account.gke_cluster_node.email}",
```

Comment:
I can't figure out why this binding is here. I think we need the binding because of this note from the upstream docs: "It is necessary to create an service account which has the role roles/iam.serviceAccountTokenCreator so it get tokens for other service accounts and can assign it to each pod."

Seems to me like this could be part of the gke-cluster module? Or the gcp-project module that creates the serviceaccount?

Perhaps I am missing something?

Maybe this one is supposed to be the k8s-snapshot SA, not the gke cluster node SA?

Reply:
So the policy above is applied to the `k8s-snapshots` service account.

The policy gives permissions to the `gke-cluster-node` SA to create tokens for the `k8s-snapshots` account. This is necessary, as the service-account-assigner runs directly on the node, therefore using the `gke-cluster-node` account when creating tokens for the individual service accounts requested by pods (in this case `k8s-snapshots` one).

Similar binding is needed for every service account to be used by pods and there'll be multiple of those in future (very soon one for the DNS challenge).

Possible alternatives:
- `gke-cluster` module  - I don't think this is a good place, as we would have to modify the module every time we add a new service/service account (The relationship between node SA and individual pods SAs is one-to-many).
- `gcp-project` - This I considered, but decided against, as imho it makes sense to keep the common module minimal. The only reson why those SA are created there is because they need permissions on the project level (that module is authoritative when setting those permissions and so the accounts need to exist when it runs).

Reply:
Makes sense, thanks.